### ISSUE TITLE
Feature request: pre-commit hook for formatter

### ISSUE THREAD

**Issue by maresb** [tamasfe/taplo#535]
It would be great if there were a [pre-commit hook](https://pre-commit.com/hooks.html) for the TOML formatter. My current go-to formatter is `pretty-format-toml`, but this suffers from a fairly serious issue that [comments are sometimes deleted](https://github.com/pappasam/toml-sort/issues/59). If this were implemented, I think it could easily become the best formatter with a pre-commit hook.

**Comment by ia0** [comment]
Thanks for the suggestion! I guess this needs both the creation of some files in this repo and a PR for the pre-commit repository. This project is currently passively maintained, so only bugs are fixed. New functionalities need a PR contribution. Feel free to open one if you get the chance.

**Comment by nikaro** [comment]
@maresb you can use mine if want: https://github.com/nikaro/taplo-pre-commit

Beware that the Docker hooks don't work on arm64 machines, until the next release of Taplo, cf. [this issue](https://github.com/nikaro/taplo-pre-commit/issues/2). The native Rust ones are fine, but the build will take a lot of time the first time you run it (after that it should be in your cache so it will be ok).

Feel free to reuse it to open a PR here if you want.

**Comment by maresb** [comment]
Thanks @nikaro, I just now got around to trying this out. My experience is that it was going very slowly during the initial install. My suspicion is that pre-commit was downloading a rust toolchain in the background, and that's why it's so slow.

I was wondering what's the purpose of your Docker hooks, but if I understood the above correctly then it saves on all the compile time. Running Docker from pre-commit seems pretty crazy to me.  I wish it could just download binaries and use those. Unfortunately I'm not familiar with rust development, so I have no sense of the challenges of doing this.

Speaking of which, there are supposed to be [binary releases](https://taplo.tamasfe.dev/cli/installation/binary.html) but it seems like the links are broken and no binary artifacts are being generated by the GitHub workflows. Should I open a separate issue for that?

**Comment by nikaro** [comment]
> Thanks @nikaro, I just now got around to trying this out. My experience is that it was going very slowly during the initial install. My suspicion is that pre-commit was downloading a rust toolchain in the background, and that's why it's so slow.

You are right that's what's happening: downloading rust toolchain and compiling. And that's why the `-docker` hook is here, to serve as a workaround to avoid having to compile (but it indeed requires Docker...).

> I wish it could just download binaries and use those. Unfortunately I'm not familiar with rust development, so I have no sense of the challenges of doing this.

There is [PR for that in pre-commit](https://github.com/pre-commit/pre-commit/pull/3020) bit it seems a bit stale lately... :-/ 

> Speaking of which, there are supposed to be [binary releases](https://taplo.tamasfe.dev/cli/installation/binary.html) but it seems like the links are broken and no binary artifacts are being generated by the GitHub workflows. Should I open a separate issue for that?

Yep, looks like the release workflow is broken: https://github.com/tamasfe/taplo/actions/runs/7742605055/workflow#L268
And it seems to be my fault 😬 the `env.RELEASE_VERSION` variable is not available in the matrix.

**Comment by mdekstrand** [comment]
As a stopgap I quickly threw together a hook that uses the binaries published to NPM: https://github.com/mdekstrand/taplo-pre-commit

**Comment by redeboer** [comment]
Since https://github.com/tamasfe/taplo/pull/549 is merged, Taplo will be published as a PyPI package on https://pypi.org/project/taplo. One could add a [`.pre-commit-hook.yaml`](https://pre-commit.com/#creating-new-hooks) to this repository, but a lightweight "mirror" (think of [ruff-pre-commit](https://github.com/astral-sh/ruff-pre-commit) or [black-pre-commit-mirror](https://github.com/psf/black-pre-commit-mirror)) would be preferable given that this repo is about 23MB. An example of the required files are at https://github.com/ComPWA/mirrors-taplo, but would be better to set up an official repo.

**Comment by maresb** [comment]
Incredible work @redeboer, I appreciate all the persistence it took to get this done!!! I have started to use https://github.com/ComPWA/mirrors-taplo as a pre-commit hook and it's working beautifully!

> but would be better to set up an official repo

:+1:

**Comment by MaxWinterstein** [comment]
@redeboer could you please just archive the project next time? Deleting it just broke our - and I guess many more - CI pipelines 😉

**Comment by maresb** [comment]
I'm pretty confused by the last two comments. 

@flying-sheep, what's wrong with autoupdate? Seems to work for me (maybe @redeboer already fixed the issue)?

Also, @MaxWinterstein, what do you mean, are you referring to https://github.com/ComPWA/taplo-pre-commit/pull/18?

Maybe these should be issues in https://github.com/ComPWA/taplo-pre-commit?

**Comment by redeboer** [comment]
> @redeboer could you please just archive the project next time? Deleting it just broke our - and I guess many more - CI pipelines 😉

Sorry about that. It was a fork that was meant for testing https://github.com/tamasfe/taplo/pull/549#issue-2123053258. It's back online now and I'll archive it.

> Maybe these should be issues in https://github.com/ComPWA/taplo-pre-commit?

Yeah probably better to post there. Once/if there is an official taplo/toml-formatter organisation (https://github.com/tamasfe/taplo/issues/403#issuecomment-1913342635) that can host this hook, I'm happy to transfer that repo there.

**Comment by flying-sheep** [comment]
> @flying-sheep, what's wrong with autoupdate?

I was responding to redeboer’s statement “[it] would be better to set up an official repo [for the pre-commit hook mirror]”

Once that exists, ComPWA/taplo-pre-commit is duplicate.

**Comment by Goldziher** [comment]
I see, sad, but you are maintaining this now @panekj ? 

If so, it would be awesome to have an official pre-commit support for this tool. It's really quite useful.

**Comment by Goldziher** [comment]
There are non-official hooks:

```yaml

  - repo: https://github.com/ComPWA/taplo-pre-commit
    rev: v0.9.3
    hooks:
      - id: taplo-format

```

But thats suboptimal. Best to have this stuff officially supported iMO.
